source/common/grpc/google_grpc_creds_impl.cc

```diff
@@ -0,0 +1,144 @@
+            .pem_root_certs = Config::DataSource::read(
+                credential.google_channel_credentials().ssl_credentials().root_certs(), true),
+            .pem_private_key = Config::DataSource::read(
+                credential.google_channel_credentials().ssl_credentials().private_key(), true),
+            .pem_cert_chain = Config::DataSource::read(
+                credential.google_channel_credentials().ssl_credentials().cert_chain(), true),
+        };
+        return grpc::SslCredentials(ssl_creds);
+      }
+    }
+    return creds;
+  }
+
+  std::string name() const override { return "default"; }
+};
+
+/**
+ * Access token implementation of Google Grpc Credentials Factory
+ * This implementation uses ssl creds for the grpc channel if available, similar to the default
+ * implementation. Additionally, it uses MetadataCredentialsFromPlugin to add a static secret to a
+ * header for call credentials. This implementation does the same thing as AccessTokenCredentials,
+ * but it's implemented as a plugin to show how a custom implementation would be created.
+ *
+ * This implementation uses the access_token field in the config to get the secret to add to the
+ * header.
+ *
+ * This can be used as an example for how to implement a more complicated custom call credentials
+ * implementation. Any blocking calls should be performed in the
+ * MetadataCredentialsFromPlugin::GetMetadata to ensure that the main thread is not blocked while
+ * initializing the channel.
+ */
+class AccessTokenGrpcCredentialsFactory : public GoogleGrpcCredentialsFactory {
+
+public:
+  std::shared_ptr<grpc::ChannelCredentials>
+  getChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config) override {
+    std::shared_ptr<grpc::ChannelCredentials> creds =
+        grpc::SslCredentials(grpc::SslCredentialsOptions());
+    std::shared_ptr<grpc::CallCredentials> call_creds = nullptr;
+    for (const auto& credential : grpc_service_config.credentials()) {
+      if (credential.has_google_channel_credentials() &&
+          credential.google_channel_credentials().has_ssl_credentials()) {
+        const grpc::SslCredentialsOptions ssl_creds = {
+            .pem_root_certs = Config::DataSource::read(
+                credential.google_channel_credentials().ssl_credentials().root_certs(), true),
+        };
+        creds = grpc::SslCredentials(ssl_creds);
+      }
+      if (credential.has_google_call_credentials() &&
+          credential.google_call_credentials().access_token() != "") {
+        call_creds = grpc::MetadataCredentialsFromPlugin(
+            std::unique_ptr<grpc::MetadataCredentialsPlugin>(new StaticHeaderAuthenticator(
+                credential.google_call_credentials().access_token())));
+      }
+    }
+    if (call_creds != nullptr) {
+      return grpc::CompositeChannelCredentials(creds, call_creds);
+    }
+    return creds;
+  }
+
+  std::string name() const override { return "access_token"; }
```

Comment:
Probably best to use standard extension namespaces here, e.g.`envoy.grpc_credentials.access_token`.